libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -775,6 +798,30 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
 {
 	setLocation(_binOp);
 
+	FunctionDefinition const* function = *_binOp.annotation().userDefinedFunction;
+	if (function)
+	{
+		_binOp.leftExpression().accept(*this);
+		_binOp.rightExpression().accept(*this);
+		setLocation(_binOp);
+
+		solAssert(function->isImplemented());
+		solAssert(function->isFree() || function->libraryFunction());
+		solAssert(function->parameters().size() == 2);
+		solAssert(function->returnParameters().size() == 1);
+		solAssert(*function->returnParameters()[0]->type() == *_binOp.annotation().type);
+
+		string left = expressionAsType(_binOp.leftExpression(), *function->parameters()[0]->type());
+		string right = expressionAsType(_binOp.rightExpression(), *function->parameters()[1]->type());
+		solAssert(!left.empty() && !right.empty());
+
+		define(_binOp) <<
+			m_context.enqueueFunctionForCodeGeneration(*function) <<
+			("(" + left + ", " + right + ")\n");
```

Comment:
This will always generate code for call internally, which is wrong for public library functions.
Same for legacy codegen actually.

Compare

```
type T is uint;

library L {
    function f(T, T) public pure returns (T) {
        assembly { mstore(0x42, 0x42) }
    }
}

using {L.f as +} for T;

contract C {
    function f(uint x, uint y) external returns(uint) {
        T a = T.wrap(x);
        T b = T.wrap(y);
        T c = L.f(a, b);
        return T.unwrap(c);
    }
```

with

```
type T is uint;

library L {
    function f(T, T) public pure returns (T) {
        assembly { mstore(0x42, 0x42) }
    }
}

using {L.f as +} for T;

contract C {
    function f(uint x, uint y) external returns(uint) {
        T a = T.wrap(x);
        T b = T.wrap(y);
        T c = a + b;
        return T.unwrap(c);
    }
```

and e.g. note the delegate call in one, but not the other, in IR output, or the link reference in one, but not the other in ``--bin`` output.

Reply:
Damn, you're right. It was already in the original PR but I did miss this in my review.

Reply:
We can hotfix this by restricting to internal library functions only for now - they make more sense anyways. There won't be enough time for fixing this properly unless we want to postpone the release.

Reply:
> Damn, you're right. It was already in the original PR but I did miss this in my review.

Yeah, also not sure why I didn't see it before now :-).